doc/reference/std/misc/atom.md

```diff
@@ -0,0 +1,339 @@
+> (atom-swap-values! a (cut * 2 <>))
+6
+12
+```
+:::
+
+## atom-reset!
+``` scheme
+(atom-reset! atom new-value) -> new-value
+```
+
+If the new value passes the atom's validator, sets the atom to it.
+Ensures that no other access is made to the atom in the middle of a reset.
+Returns the new value.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42))
+> (atom-reset! a 23)
+23
+> (atom-deref a)
+23
+```
+:::
+
+## atom-reset-values!
+``` scheme
+(atom-reset-values! atom new-value) -> old-value new-value
+```
+
+Similar to `atom-reset!` but returns two values, the old and the new.
+Similar to Clojure `atom-reset-vals!` but returns values instead of a vector.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42))
+> (atom-reset-values! a 23)
+42
+23
+> (atom-deref a)
+23
+```
+:::
+
+## atom-compare-and-set!
+
+``` scheme
+(atom-compare-and-set! atom oldval newval) -> bool
+```
+
+Atomically sets the value of atom to newval if and only if the
+current value of the atom is identical to oldval.
+Returns `#t` if set happened, else `#f`.
+Similar to Clojure `compare-and-set!`.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42))
+> (atom-compare-and-set! a 42 23)
+#t
+> (atom-deref a)
+23
+> (atom-compare-and-set! a 42 24)
+#f
+> (atom-deref a)
+23
+```
+:::
+
+## atom-validator
+
+``` scheme
+(atom-validator atom) -> validator
+```
+
+Access the current atom's validator, a function that takes a new value as argument
+and returns a true value iff the atom will accept that value.
+The value `#f` is a stand-in for not doing any validation,
+which is equivalent to the function `true` that accepts every value.
+Similar to Clojure `get-validator`.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42 validator: number?))
+> (atom-reset! a 1)
+1
+> (atom-reset! a "foo")
+*** ERROR IN with-lock --
+*** ERROR IN "misc/atom.ss"@48.29 [BadArgument]: bad argument; expected valid atom value
+--- irritants: update-atom! "foo"
+```
+:::
+
+## atom-validator-set!
+
+``` scheme
+(atom-validator-set! atom validator) -> _
+```
+
+Modify the current atom's validator, a function that takes a new value as argument
+and returns a true value iff the atom will accept that value.
+The value `#f` is a stand-in for not doing any validation,
+which is equivalent to the function `true` that accepts every value.
+The current value is grandfathered and not validated;
+the validator is not validated for whether it accepts the current value.
+The function is not guaranteed to return any specific value.
+Similar to Clojure `set-validator!`.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42 validator: number?))
+> (atom-reset! a 1)
+1
+> (atom-validator-set! a string?)
+> (atom-reset! a "foo")
+"foo"
+> (atom-reset! a 2)
+*** ERROR IN "misc/atom.ss"@48.29 [BadArgument]: bad argument; expected valid atom value
+--- irritants: update-atom! 1
+```
+:::
+
+## atom-add-watch!
+
+``` scheme
+(atom-add-watch! atom key function) -> void
+```
+
+Adds a watch function to an atom.
```

Comment:
``` scheme
(atom-add-watch! atom key watch-function) -> void
```

Adds a watch function `watch-function` to `atom`. The watch function must take 4 arguments: a `key`, the `atom`, its `old-state` and its `new-state`. Whenever the `atom`'s state is changed, after the new value has been validated, each of the registered watch functions will be called in an unspecified order while the atom lock is held. No changes to the `atom` are allowed in the watch functions themselves, though they may send asynchronous messages or spawn threads that will cause a cascade of further changes. Watch `key`s must be unique per `atom` and can be used to remove the watch function iwth `atom-remove-watch!`. They are otherwise considered opaque by the watch mechanism.

It is unspecified whether the state has been modified yet, so use `old-state` and `new-state` rather than `atom-deref`. Note that watch functions may be called from multiple threads simultaneously. (In the current implementation, the state will already have been modified and no threads will be spawned.)

Similar to Clojure's `add-watch` function.